File: JS/strings/string_manipulation.js
// ...existing code...
/**
 * Reverse a string.
 * @param {string|null|undefined} input
 * @returns {string|null}
 */
function reverseString(input) {
  if (input == null) return null;
  return String(input).split('').reverse().join('');
}

/**
 * Check if a string is a palindrome (ignores whitespace and case).
 * @param {string|null|undefined} s
 * @returns {boolean}
 */
function isPalindrome(s) {
  if (s == null) return false;
  const clean = String(s).replace(/\s+/g, '').toLowerCase();
  return clean === clean.split('').reverse().join('');
}

/**
 * Character frequency in a string — returns a Map<char, count>.
 * @param {string|null|undefined} s
 * @returns {Map<string, number>}
 */
function charFrequency(s) {
  const freq = new Map();
  if (s == null) return freq;
  for (const ch of String(s)) {
    freq.set(ch, (freq.get(ch) || 0) + 1);
  }
  return freq;
}

/**
 * Remove duplicates from array while preserving order.
 * @param {Array|null|undefined} arr
 * @returns {Array}
 */
function removeDuplicatesFromArray(arr) {
  if (arr == null) return [];
  return Array.from(new Set(arr));
}

/**
 * Find the second largest distinct number in an array.
 * Returns null if there isn't one.
 * @param {number[]|null|undefined} arr
 * @returns {number|null}
 */
function secondLargest(arr) {
  if (!Array.isArray(arr) || arr.length < 2) return null;
  let max = -Infinity;
  let second = -Infinity;
  for (const n of arr) {
    if (typeof n !== 'number' || Number.isNaN(n)) continue;
    if (n > max) {
      second = max;
      max = n;
    } else if (n < max && n > second) {
      second = n;
    }
  }
  return second === -Infinity ? null : second;
}

/**
 * Check if a string contains only digits.
 * @param {string|null|undefined} s
 * @returns {boolean}
 */
function isNumeric(s) {
  if (s == null) return false;
  return /^[0-9]+$/.test(String(s));
}

/**
 * Reverse the words in a sentence (preserve single spaces between words).
 * @param {string|null|undefined} sentence
 * @returns {string|null}
 */
function reverseWords(sentence) {
  if (sentence == null) return null;
  return String(sentence).trim().split(/\s+/).reverse().join(' ');
}

/**
 * First non-repeating character in a string. Returns the character or null.
 * @param {string|null|undefined} s
 * @returns {string|null}
 */
function firstNonRepeatingChar(s) {
  if (s == null) return null;
  const count = new Map();
  for (const ch of String(s)) {
    count.set(ch, (count.get(ch) || 0) + 1);
  }
  for (const ch of String(s)) {
    if (count.get(ch) === 1) return ch;
  }
  return null;
}

/**
 * Sum array using reduce. Non-number entries are coerced by +
 * @param {number[]|null|undefined} arr
 * @returns {number}
 */
function sumArray(arr) {
  if (!Array.isArray(arr)) return 0;
  return arr.reduce((acc, v) => acc + (typeof v === 'number' ? v : Number(v) || 0), 0);
}

/**
 * Check if array contains duplicates.
 * @param {Array|null|undefined} arr
 * @returns {boolean}
 */
function hasDuplicates(arr) {
  if (!Array.isArray(arr)) return false;
  const seen = new Set();
  for (const v of arr) {
    if (seen.has(v)) return true;
    seen.add(v);
  }
  return false;
}

module.exports = {
  reverseString,
  isPalindrome,
  charFrequency,
  removeDuplicatesFromArray,
  secondLargest,
  isNumeric,
  reverseWords,
  firstNonRepeatingChar,
  sumArray,
  hasDuplicates,
};

File: Python/strings/string_manipulation.py
"""
String and array utility functions.
Mirrors JS implementations with None handling where applicable.
"""
from typing import Optional, List, Dict, Set
import re


def reverse_string(input_str: Optional[str]) -> Optional[str]:
    """
    Reverse a string.
    
    Args:
        input_str: string or None
    
    Returns:
        reversed string or None if input is None
    """
    if input_str is None:
        return None
    return input_str[::-1]


def is_palindrome(s: Optional[str]) -> bool:
    """
    Check if a string is a palindrome (ignores whitespace and case).
    
    Args:
        s: string or None
    
    Returns:
        True if palindrome, False otherwise (None returns False)
    """
    if s is None:
        return False
    clean = re.sub(r'\s+', '', s).lower()
    return clean == clean[::-1]


def char_frequency(s: Optional[str]) -> Dict[str, int]:
    """
    Return character frequency count in a string.
    
    Args:
        s: string or None
    
    Returns:
        Dict with character -> count mapping
    """
    freq = {}
    if s is None:
        return freq
    for ch in s:
        freq[ch] = freq.get(ch, 0) + 1
    return freq


def remove_duplicates_from_array(arr: Optional[List]) -> List:
    """
    Remove duplicates from array while preserving order.
    
    Args:
        arr: list or None
    
    Returns:
        list with duplicates removed
    """
    if arr is None:
        return []
    seen = set()
    result = []
    for item in arr:
        if item not in seen:
            seen.add(item)
            result.append(item)
    return result


def second_largest(arr: Optional[List[int]]) -> Optional[int]:
    """
    Find the second largest distinct number in an array.
    
    Args:
        arr: list of numbers or None
    
    Returns:
        second largest number or None if not found
    """
    if not isinstance(arr, list) or len(arr) < 2:
        return None
    
    distinct = list(set(arr))
    if len(distinct) < 2:
        return None
    
    distinct.sort(reverse=True)
    return distinct[1]


def is_numeric(s: Optional[str]) -> bool:
    """
    Check if a string contains only digits.
    
    Args:
        s: string or None
    
    Returns:
        True if all characters are digits, False otherwise
    """
    if s is None:
        return False
    return bool(re.match(r'^[0-9]+$', s))


def reverse_words(sentence: Optional[str]) -> Optional[str]:
    """
    Reverse the words in a sentence (preserve single spaces between words).
    
    Args:
        sentence: string or None
    
    Returns:
        string with reversed word order or None if input is None
    """
    if sentence is None:
        return None
    words = sentence.strip().split()
    return ' '.join(reversed(words))


def first_non_repeating_char(s: Optional[str]) -> Optional[str]:
    """
    Find the first non-repeating character in a string.
    
    Args:
        s: string or None
    
    Returns:
        first non-repeating character or None if not found
    """
    if s is None:
        return None
    
    count = {}
    for ch in s:
        count[ch] = count.get(ch, 0) + 1
    
    for ch in s:
        if count[ch] == 1:
            return ch
    
    return None


def sum_array(arr: Optional[List]) -> float:
    """
    Sum array of numbers. Non-numeric entries are coerced to 0.
    
    Args:
        arr: list of numbers or None
    
    Returns:
        sum of array elements, 0 if None or empty
    """
    if not isinstance(arr, list):
        return 0
    
    total = 0.0
    for v in arr:
        try:
            total += float(v) if isinstance(v, (int, float)) else 0
        except (ValueError, TypeError):
            pass
    return total


def has_duplicates(arr: Optional[List]) -> bool:
    """
    Check if array contains duplicate elements.
    
    Args:
        arr: list or None
    
    Returns:
        True if duplicates found, False otherwise
    """
    if not isinstance(arr, list):
        return False
    return len(arr) != len(set(arr))
